core/src/apps/cardano/sign_tx.py

```diff
@@ -226,11 +227,43 @@ def _validate_outputs(
             )
 
         _validate_token_bundle(output.token_bundle)
+        _validate_max_tx_output_size(keychain, output, protocol_magic, network_id)
 
     if total_amount > LOVELACE_MAX_SUPPLY:
         raise wire.ProcessError("Total transaction amount is out of range!")
 
 
+def _validate_max_tx_output_size(
+    keychain: seed.Keychain,
+    output: CardanoTxOutputType,
+    protocol_magic: int,
+    network_id: int,
+) -> None:
+    """
+    Output size is currently limited to 4000 bytes at protocol level. Given
+    the maximum transaction size Trezor can handle (~9kB), we also want to enforce
+    this size limit here so that when the limit is raised at protocol level again,
+    Trezor would still not be able to produce larger outputs than it could
+    reliably spend. Once Cardano-transaction signing is refactored to be completely
+    streamed and maximum supported transaction size is thus raised, this limit can be lifted.
+    """
+    cborized_output = _cborize_output(keychain, output, protocol_magic, network_id)
+
+    assert len(cborized_output) == 2
+    # only the output value is used for counting the size in cardano-ledger-specs
+    cborized_output_value = cborized_output[1]
+    serialized_output_chunks = cbor.encode_streamed(cborized_output_value)
+
+    serialized_output_size = 0
+    for chunk in serialized_output_chunks:
+        serialized_output_size += len(chunk)
+
+    if serialized_output_size > MAX_TX_OUTPUT_SIZE:
+        raise wire.ProcessError(
+            "Maximum tx output size (%s bytes) exceeded!" % MAX_TX_OUTPUT_SIZE
```

Comment:
I'd update

```suggestion
            "Maximum tx output size (%s bytes) exceeded!" % MAX_TX_OUTPUT_SIZE
```
to

```suggestion
            "Maximum tx output value size (%s bytes) exceeded!" % MAX_TX_OUTPUT_SIZE
```

to reflect reality